firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/PacketReceiver.cpp

```diff
@@ -33,214 +33,93 @@ namespace Pufferfish::Driver::Serial::Nonin {
-  static const uint8_t mask_start_of_frame = 0x80;
-  static const uint8_t mask_snsd = 0x40;
-  static const uint8_t mask_artf = 0x20;
-  static const uint8_t mask_oot = 0x10;
-  static const uint8_t mask_snsa = 0x08;
-  static const uint8_t mask_yprf = 0x06;
-  static const uint8_t mask_rprf = 0x04;
-  static const uint8_t mask_gprf = 0x02;
-
-  /* BIT7: Always high */
-  sensor_measurements.bit7[frame_index] = (byte_value & mask_start_of_frame) == 0x00;
-  /* BIT6: SNSD: Sensor Disconnect – Sensor is not connected to oximeter or
-   * sensor is inoperable */
-  sensor_measurements.sensor_disconnect[frame_index] = (byte_value & mask_snsd) != 0x00;
-  /* BIT5: ARTF: Artifact – A detected pulse beat didn’t match the current pulse
-   * interval */
-  sensor_measurements.artifact[frame_index] = (byte_value & mask_artf) != 0x00;
-  /* BIT4: OOT: Out Of Track – An absence of consecutive good pulse signals */
-  sensor_measurements.out_of_track[frame_index] = (byte_value & mask_oot) != 0x00;
-  /* BIT3: SNSA: Sensor Alarm – Sensor is providing unusable data for analysis
-   */
-  sensor_measurements.sensor_alarm[frame_index] = (byte_value & mask_snsa) != 0x00;
-
-  /**
-   * BIT2 and BIT1: YPRF: Yellow Perfusion – Amplitude representation of medium
-   * signal quality.
-   * BIT2: RPRF: Red Perfusion – Amplitude representation of low
-   * signal quality.
-   * BIT1: GPRF: Green Perfusion – Amplitude representation of
-   * high signal quality.
-   */
-  if ((byte_value & mask_yprf) == mask_yprf) {
-    sensor_measurements.signal_perfusion[frame_index] = SignalAmplitude::yellow_perfusion;
-  } else if ((byte_value & mask_rprf) == mask_rprf) {
-    sensor_measurements.signal_perfusion[frame_index] = SignalAmplitude::red_perfusion;
-  } else if ((byte_value & mask_gprf) == mask_gprf) {
-    sensor_measurements.signal_perfusion[frame_index] = SignalAmplitude::green_perfusion;
+    Sample &sensor_measurements, const size_t &frame_index, const uint8_t &byte_value) {
+  sensor_measurements.sensor_disconnect[frame_index] = (byte_value & snsd) != 0x00;
+  sensor_measurements.artifact[frame_index] = (byte_value & artf) != 0x00;
+  sensor_measurements.out_of_track[frame_index] = (byte_value & oot) != 0x00;
+  sensor_measurements.sensor_alarm[frame_index] = (byte_value & snsa) != 0x00;
+  sensor_measurements.bit7[frame_index] = (byte_value & start_of_frame) == 0x00;
+
+  if ((byte_value & yprf) == yprf) {
+    sensor_measurements.signal_perfusion[frame_index] = SignalQuality::yellow_perfusion;
+  } else if ((byte_value & rprf) == rprf) {
+    sensor_measurements.signal_perfusion[frame_index] = SignalQuality::red_perfusion;
+  } else if ((byte_value & gprf) == gprf) {
+    sensor_measurements.signal_perfusion[frame_index] = SignalQuality::green_perfusion;
   } else {
-    sensor_measurements.signal_perfusion[frame_index] = SignalAmplitude::no_perfusion;
+    sensor_measurements.signal_perfusion[frame_index] = SignalQuality::no_perfusion;
   }
```

Comment:
[for a future PR] To avoid copying `sensor_measurements.signal_perfusion[frame_index]` so many times, maybe we could first get a non-const ref as `SignalPerfusion &signal_perfusion = sensor_measurements.signal_perfusion[frame_index]` and then just assign to `signal_perfusion`.